feat(profiling): add nodejs profiling docs (#5776)
getsentry/sentry-docs: src/platforms/node/common/profiling/index.mdx

<Note>

Node.js profiling is currently in alpha, which means that it's still in progress and may have bugs. We recognize the irony. If you have any questions or feedback, please email us at [profiling@sentry.io](mailto:profiling@sentry.io). If you’re adopting Profiling in a high-throughput environment, we recommend testing prior to deployment to ensure that your service’s performance characteristics maintain expectations.

</Note>

```javascript
const Sentry = require("@sentry/node");
// Note: You MUST import @sentry/tracing package before @sentry/profiling-node
require("@sentry/tracing");
const {ProfilingIntegration} = require("@sentry/profiling-node");

Sentry.init({
  dsn: "___PUBLIC_DSN___",
  integrations: [
    // Add our Profilling integration
    new ProfilingIntegration(),
  ],
  tracesSampleRate: 1.0,
  // Set sampling rate for profiling
  profilesSampleRate: 1.0,
});

const transaction = Sentry.startTransaction({
  op: "transaction",
  name: "My Transaction",
});

// Any code executed between startTransaction and transaction.finish
// will now be automatically profiled.

transaction.finish();
```

> 

Reviewer: Removing extra line break
```suggestion
```

## How Does It Work?

> Under the hood, the sentry profiler leverages V8's [CpuProfiler](https://v8docs.nodesource.com/node-18.2/d2/d34/classv8_1_1_cpu_profiler.html) to collect stack samples. This means that sentry/profiling-node is written as a [native add-on](https://nodejs.org/docs/latest-v18.x/api/addons.html) for Node and won't run in environments like Deno or Bun. Profiling enhances tracing by providing profiles for individual transactions. This allows you to look at higher level performance information like transaction and span durations before diving deeper and looking at profiles.

Reviewer: Should "Sentry" in the first sentence be capitalized? I get why it isn't in the second sentence when it's part of the a larger term. I changed "leverages" to "uses" to try to simplify the language as much as possible.

```suggestion
Under the hood, the Sentry profiler uses V8's [CpuProfiler](https://v8docs.nodesource.com/node-18.2/d2/d34/classv8_1_1_cpu_profiler.html) to collect stack samples. This means that sentry/profiling-node is written as a [native add-on](https://nodejs.org/docs/latest-v18.x/api/addons.html) for Node and won't run in environments like Deno or Bun. Profiling enhances tracing by providing profiles for individual transactions. This allows you to look at higher level performance information like transaction and span durations before diving deeper and looking at profiles.
```

## Runtime Flags

> The Profiling package enables you to run the profiler in either [eager or lazy mode](https://v8docs.nodesource.com/node-18.2/d2/dc3/namespacev8.html#a874b4921ddee43bef58d8538e3149374). During initial tests, when we started a profile while no other profiles were running, we saw that the default profiling mode (lazy) sometimes caused startProfiling calls to take a long time, (sometimes in the range of a couple hundred milliseconds). The benefit of lazy logging is that it doesn't add any overhead when profiles aren't running.

Reviewer: Just wondering if we can make this more concise:
```suggestion
The **Profiling** package enables you to run the profiler in either [eager or lazy mode](https://v8docs.nodesource.com/node-18.2/d2/dc3/namespacev8.html#a874b4921ddee43bef58d8538e3149374). When a profile is started while no other profiles are running, the default profiling mode (lazy) sometimes causes `startProfiling` calls to take a long time (sometimes in the range of a couple hundred milliseconds). The benefit of lazy logging is that it doesn't add any overhead when profiles aren't running.
```

> ## Runtime flags
> 
> Profiling package enables you to run the profiler in either eager or lazy [mode](https://v8docs.nodesource.com/node-18.2/d2/dc3/namespacev8.html#a874b4921ddee43bef58d8538e3149374). During our initial tests, we have seen that the default profiling mode (lazy) can cause calls to startProfiling to take a long time - sometimes in the range of couple hundred milliseconds. We found that this occurs when we are starting a profile when no other profiles are started. This is the benefit of lazy logging, it adds no overhead when no profiles are running.
> 

Reviewer 2: ```suggestion
## Runtime Flags

The Profiling package enables you to run the profiler in either [eager or lazy mode](https://v8docs.nodesource.com/node-18.2/d2/dc3/namespacev8.html#a874b4921ddee43bef58d8538e3149374). During initial tests, when we started a profile while no other profiles were running, we saw that the default profiling mode (lazy) sometimes caused startProfiling calls to take a long time, (sometimes in the range of a couple hundred milliseconds). The benefit of lazy logging is that it doesn't add any overhead when profiles aren't running.

```
Hoping that my edits didn't change the meaning of what you were trying to say :)

> If this behavior is prohibitive for your use case, there's an environment flag that you can set which will cause the profiler to run in eager mode. Eager mode improves the performance of calls to startProfiling, but the tradeoff is small, constant CPU overhead.

Reviewer: ```suggestion
If this behavior is prohibitive for your use case, you can set the environment flag shown below, which will cause the profiler to run in eager mode. Eager mode improves the performance of calls to `startProfiling`, but the tradeoff is small, constant CPU overhead.
```

> If you find this behavior prohibitive for your use case, there is an environment flag that you can set which will cause the profiler to run in eager mode. Eager mode improves the performance of calls to startProfiling at the tradeoff of small constant CPU overhead.
> 

Reviewer 2: ```suggestion
If this behavior is prohibitive for your use case, there's an environment flag that you can set which will cause the profiler to run in eager mode. Eager mode improves the performance of calls to startProfiling, but the tradeoff is small, constant CPU overhead.

```

```bash
// Run profiler in eager mode
SENTRY_PROFILER_LOGGING_MODE=eager node script.js
```

> It's always advisable that you have your own CPU resource-monitoring in place, because the actual resource use could be environment-dependent.

Reviewer: ```suggestion
We recommend you have your own CPU resource-monitoring in place, because the actual resource use could be environment-dependent.
```

> It is always advised that you have your own CPU resource monitoring in place as the actual resource use could be environment dependant.

Reviewer 2: ```suggestion
It's always advisable that you have your own CPU resource-monitoring in place, because the actual resource use could be environment-dependent.
```

> If you have feedback or experience different behavior please open an issue on sentry/profiling-node repository.

Reviewer: ```suggestion
If you have feedback or experience different behavior, please open an issue on sentry/profiling-node repository.
```
